autodiff/forward/real/real.hpp

```diff
@@ -0,0 +1,835 @@
+            a[i] = x[i] - Sum<1, i>([&](auto j) constexpr {
+                constexpr auto c = BinomialCoefficient<i.index - 1, j.index - 1>;
+                return c * x[i - j] * a[j];
+            });
+            a[i] /= x[0];
+
+            res[i] = 0.5 * Sum<0, i>([&](auto j) constexpr {
+                constexpr auto c = BinomialCoefficient<i.index - 1, j.index>;
+                return c * a[i - j] * res[j];
+            });
+        });
+    }
+    return res;
+}
+
+template<size_t N, typename T>
+constexpr auto pow(const Real<N, T>& x, const Real<N, T>& y)
+{
+    Real<N, T> res;
+    res[0] = pow(x[0], y[0]);
+    if constexpr (N > 0)
+    {
+        assert(x[0] != 0 && "autodiff: pow(x, y) has undefined derivatives when x = 0");
+        Real<N, T> lnx = log(x);
+        Real<N, T> a;
+        For<1, N + 1>([&](auto i) constexpr {
+            a[i] = Sum<0, i + 1>([&](auto j) constexpr {
+                constexpr auto c = BinomialCoefficient<i.index, j.index>;
+                return c * y[i - j] * lnx[j];
+            });
+
+            res[i] = Sum<0, i>([&](auto j) constexpr {
+                constexpr auto c = BinomialCoefficient<i.index - 1, j.index>;
+                return c * a[i - j] * res[j];
+            });
+        });
+    }
+    return res;
+}
+
+template<size_t N, typename T, typename U, EnableIf<isNumber<U>>...>
+constexpr auto pow(const Real<N, T>& x, const U& c)
+{
+    Real<N, T> res;
+    res[0] = pow(x[0], static_cast<T>(c));
+    if constexpr (N > 0)
+    {
+        assert(x[0] != 0 && "autodiff: pow(x, y) has undefined derivatives when x = 0");
+        Real<N, T> a = c * log(x);
+        For<1, N + 1>([&](auto i) constexpr {
+            res[i] = Sum<0, i>([&](auto j) constexpr {
+                constexpr auto c = BinomialCoefficient<i.index - 1, j.index>;
+                return c * a[i - j] * res[j];
+            });
+        });
+    }
+    return res;
+}
+
+template<size_t N, typename T, typename U, EnableIf<isNumber<U>>...>
+constexpr auto pow(const U& c, const Real<N, T>& y)
+{
+    Real<N, T> res;
+    res[0] = pow(static_cast<T>(c), y[0]);
+    if constexpr (N > 0)
+    {
+        assert(c != 0 && "autodiff: pow(x, y) has undefined derivatives when x = 0");
+        Real<N, T> a = y * log(c);
+        For<1, N + 1>([&](auto i) constexpr {
+            res[i] = Sum<0, i>([&](auto j) constexpr {
+                constexpr auto c = BinomialCoefficient<i.index - 1, j.index>;
+                return c * a[i - j] * res[j];
+            });
+        });
+    }
+    return res;
+}
+
+//=====================================================================================================================
+//
+// TRIGONOMETRIC FUNCTIONS
+//
+//=====================================================================================================================
+
+template<size_t N, typename T>
+auto sincos(const Real<N, T>& x) -> std::tuple<Real<N, T>, Real<N, T>>
```

Comment:
Can we move it to a detail namespace? Since it is not good for the user.

Reply:
I think we now have 99% of all codes in a `detail` namespace and whatever is supposed to be public, we export to the parent namespace `autodiff`. Hopefully we'll be able to use modules in the next C++ version, C++20, which will make this less painful.